flaml/searcher/search_thread.py

```diff
@@ -0,0 +1,132 @@
+'''!
+ * Copyright (c) 2020-2021 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the
+ * project root for license information.
+'''
+from typing import Dict, Optional
+import numpy as np
+try:
+    from ray.tune.suggest import Searcher
+except ImportError:
+    from .suggestion import Searcher
+from .flow2 import FLOW2
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+class SearchThread:
+    '''Class of global or local search thread
+    '''
+
+    cost_attr = 'time_total_s'
+
+    def __init__(self, mode: str = "min", 
+                 search_alg: Optional[Searcher] = None):
+        ''' When search_alg is omitted, use local search FLOW2
+        '''
+        self._search_alg = search_alg
+        self._mode = mode
+        self._metric_op = 1 if mode=='min' else -1
+        self.cost_best = self.cost_last = self.cost_total = self.cost_best1 = \
+            getattr(search_alg, 'cost_incumbent', 0)
+        self.cost_best2 = 0
+        self.obj_best1 = self.obj_best2 = getattr(
+            search_alg, 'best_obj', np.inf) # inherently minimize        
+        # eci: expected cost for improvement
+        self.eci = self.cost_best
+        self.priority = self.speed = 0
+        
+    def suggest(self, trial_id: str) -> Optional[Dict]:
+        ''' use the suggest() of the underlying search algorithm
+        '''
+        if isinstance(self._search_alg, FLOW2):
+            config = self._search_alg.suggest(trial_id)
+        else:
+            try:
+                config = self._search_alg.suggest(trial_id)
+            except:
+                logger.warning(
```

Comment:
what types of errors can happen? Will it still happen for the next iteration? Can we make sure it does not get stuck?